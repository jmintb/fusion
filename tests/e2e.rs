use anyhow::Result;
use assert_cmd::Command;
use rstest::rstest;
use std::path::PathBuf;

#[rstest]
fn test_well_formed_programs(#[files("./test_programs/*.ts")] path: PathBuf) -> Result<()> {
    let mut cmd = Command::cargo_bin("typed_script")?;
    let assert = cmd.arg("run").arg(path.clone()).assert();
    assert.success();
    insta::assert_snapshot!(
        format!(
            "test_well_formed_programs_{}",
            path.file_name().unwrap().to_str().unwrap()
        ),
        String::from_utf8(cmd.ok().unwrap().stdout)?
    );
    Ok(())
}

#[rstest]
fn snapshop_mlir_output(#[files("./test_programs/*.ts")] path: PathBuf) -> Result<()> {
    let mut cmd = Command::cargo_bin("typed_script")?;
    let assert = cmd
        .args(&["build", "--emit-mlir"])
        .arg(path.clone())
        .assert();
    assert.success();
    insta::assert_snapshot!(
        format!(
            "test_mlir_snapshot_{}",
            path.file_name().unwrap().to_str().unwrap()
        ),
        String::from_utf8(cmd.ok().unwrap().stdout)?
    );
    Ok(())
}

#[rstest]
fn snapshop_ast_output(#[files("./test_programs/*.ts")] path: PathBuf) -> Result<()> {
    let mut cmd = Command::cargo_bin("typed_script")?;
    let assert = cmd
        .args(&["build", "--emit-ast"])
        .arg(path.clone())
        .assert();
    assert.success();
    insta::assert_snapshot!(
        format!(
            "test_ast_snapshot_{}",
            path.file_name().unwrap().to_str().unwrap()
        ),
        String::from_utf8(cmd.ok().unwrap().stdout)?
    );
    Ok(())
}
