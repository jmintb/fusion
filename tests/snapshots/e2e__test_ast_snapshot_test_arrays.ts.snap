---
source: tests/e2e.rs
expression: "String::from_utf8(cmd.ok().unwrap().stdout)?"
---
[
    Decl(
        Function {
            keywords: [
                Normal,
            ],
            id: TSIdentifier(
                "main",
            ),
            arguments: [],
            body: Some(
                [
                    Assignment(
                        Assignment {
                            id: TSIdentifier(
                                "val",
                            ),
                            expression: Value(
                                String(
                                    "test",
                                ),
                                String,
                            ),
                            typed_annotation: None,
                        },
                    ),
                    Assignment(
                        Assignment {
                            id: TSIdentifier(
                                "vals",
                            ),
                            expression: Array(
                                Array {
                                    item_type: String,
                                    items: [
                                        Value(
                                            String(
                                                "first \\n",
                                            ),
                                            String,
                                        ),
                                        Value(
                                            String(
                                                "second \\n",
                                            ),
                                            String,
                                        ),
                                        Value(
                                            String(
                                                "third \\n",
                                            ),
                                            String,
                                        ),
                                    ],
                                },
                            ),
                            typed_annotation: None,
                        },
                    ),
                    Expression(
                        Call(
                            TSIdentifier(
                                "print",
                            ),
                            [
                                ArrayLookup(
                                    ArrayLookup {
                                        array_identifier: TSIdentifier(
                                            "vals",
                                        ),
                                        index_expression: Value(
                                            Integer(
                                                0,
                                            ),
                                            Integer,
                                        ),
                                    },
                                ),
                                Value(
                                    Integer(
                                        7,
                                    ),
                                    Integer,
                                ),
                            ],
                        ),
                    ),
                    Expression(
                        Call(
                            TSIdentifier(
                                "print",
                            ),
                            [
                                ArrayLookup(
                                    ArrayLookup {
                                        array_identifier: TSIdentifier(
                                            "vals",
                                        ),
                                        index_expression: Value(
                                            Integer(
                                                1,
                                            ),
                                            Integer,
                                        ),
                                    },
                                ),
                                Value(
                                    Integer(
                                        8,
                                    ),
                                    Integer,
                                ),
                            ],
                        ),
                    ),
                    Expression(
                        Call(
                            TSIdentifier(
                                "print",
                            ),
                            [
                                ArrayLookup(
                                    ArrayLookup {
                                        array_identifier: TSIdentifier(
                                            "vals",
                                        ),
                                        index_expression: Value(
                                            Integer(
                                                2,
                                            ),
                                            Integer,
                                        ),
                                    },
                                ),
                                Value(
                                    Integer(
                                        7,
                                    ),
                                    Integer,
                                ),
                            ],
                        ),
                    ),
                    Expression(
                        Return(
                            Return {
                                expression: None,
                            },
                        ),
                    ),
                ],
            ),
            return_type: Unit,
        },
    ),
]
